src/compiler/emitter.ts

```diff
@@ -480,6 +494,14 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
                 emit(sourceFile);
             }
 
+            function emitConcatenatedModule(sourceFile: SourceFile): void {
+                currentSourceFile = sourceFile;
+                exportFunctionForFile = undefined;
+                let canonicalName = getExternalModuleNameFromPath(host, sourceFile.fileName);
+                sourceFile.moduleName = sourceFile.moduleName || canonicalName;

```

Comment:
i would capture this in a variable instead of mutating it. the tree should be immutable post parsing (modulo intentional breaks to the previous rule :D)